### PR TITLE
[Fix #11802] Improve handling of [] and () with percent symbol arrays

### LINKED ISSUE #11802: Symbol/WordArray style "percent" should also consider [ ] chars as "bad"
The current behavior of the Symbol/WordArray cop's default style ("percent") is as follows:

```ruby
# bad
[:foo, :[]=]

# bad (contains spaces)
%w[foo\ bar baz\ quux]

# good (allows [ ] chars)
%i[foo \[\]=]
```

I think we should change this so that `[ ]` trigger the same special handling that space char does, and the preferred style here would be `[:foo, :[]=]`

### PR DESCRIPTION
**Pull request by jasondoc3**
Addresses https://github.com/rubocop/rubocop/issues/11802, but also accounts for `()`

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/internal_affairs/undefined_config.rb [hunk 1]
```diff
@@ -8,7 +8,7 @@ class UndefinedConfig < Base
         ALLOWED_CONFIGURATIONS = %w[
           Safe SafeAutoCorrect AutoCorrect Severity StyleGuide Details Reference Include Exclude
         ].freeze
-        RESTRICT_ON_SEND = %i[[] fetch].freeze
+        RESTRICT_ON_SEND = [:[], :fetch].freeze
```
Comment (koic) [inline lib/rubocop/cop/internal_affairs/undefined_config.rb:11]: If bracket do not contain spaces, they should not be changed.

So, only the following case that need escaping can be changed from percent literal style to bracket style:

```ruby
%i[[ ]] #=> [:"[", :"]"]
```

This case where brackets do not contain spaces should not be changed:

```ruby
%i[[]]  #=> [:[]]
```

That is, it must be checked for spaces in brackets. Can you update the entire changes in this PR?

Comment (jasondoc3) [reply lib/rubocop/cop/internal_affairs/undefined_config.rb:11]: @koic I think that makes sense to me!

But one thing that I notice is that while `%i[[] one]` does not currently raise an offense, rubocop still autocorrects `[:[], :one]` to `%i[\[\] one]`, which feels inconsistent 🤔 Any thoughts?

Comment (koic) [reply lib/rubocop/cop/internal_affairs/undefined_config.rb:11]: When `EnforcedStyle: percent`, `%i[[] one]` (no spaces in the brackets) will need to adjust the implementation if necessary to be accepted.

File: lib/rubocop/cop/style/symbol_array.rb [hunk 2]
```diff
@@ -22,6 +22,15 @@ module Style
       #   # bad
       #   [:foo, :bar, :baz]
       #
+      #   # bad (contains spaces)
+      #   %i[foo\ bar baz\ quux]
+      #
+      #   # bad (contains [] with spaces)
+      #   %i[foo\ \[ \]]
```
Comment (bbatsov) [inline lib/rubocop/cop/style/symbol_array.rb:29]: I'd remove the escaped space after `foo` to make this example clearer.

File: lib/rubocop/cop/style/symbol_array.rb [hunk 3]
```diff
@@ -22,6 +22,15 @@ module Style
...
+      #
+      #   # bad (contains () with spaces)
+      #   %i(foo\ \( \))
```
Comment (bbatsov) [inline lib/rubocop/cop/style/symbol_array.rb:32]: I'd remove the escaped space after `foo` to make this example clearer.

**Comment by bbatsov** [comment]
@jasondoc3 ping :-)

**Comment by jasondoc3** [comment]
@bbatsov Removed the escaped spaces! I _think_ I addressed the concerns of @koic, but if not I can give another look :eyes:

File: lib/rubocop/cop/style/symbol_array.rb [hunk 4]
```diff
@@ -57,13 +66,24 @@ def on_array(node)
 
         private
 
-        def symbols_contain_spaces?(node)
+        def complex_content?(node)
+          delimiters = ['[', ']', '(', ')']
```
Comment (koic) [inline lib/rubocop/cop/style/symbol_array.rb:70]: @jasondoc3 It could be a constant instead of a variable :-)

Comment (jasondoc3) [reply lib/rubocop/cop/style/symbol_array.rb]: Done!

**Comment by koic** [comment]
Thanks!

**Pull request merged**